src/Features/Core/Portable/IntroduceUsingStatement/AbstractIntroduceUsingStatementCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,284 @@
+                return default;
+            }
+
+            return (declarationSyntax, declarator.Symbol);
+        }
+
+        private static bool IsLegalUsingStatementType(Compilation compilation, ITypeSymbol disposableType, ITypeSymbol type)
+        {
+            if (disposableType == null)
+            {
+                return false;
+            }
+
+            // CS1674: type used in a using statement must be implicitly convertible to 'System.IDisposable'
+            return compilation.ClassifyCommonConversion(type, disposableType).IsImplicit;
+        }
+
+        private async Task<Document> IntroduceUsingStatementAsync(
+            Document document,
+            TextSpan span,
+            CancellationToken cancellationToken)
+        {
+            var (declarationStatement, localVariable) = await FindDisposableLocalDeclaration(document, span, cancellationToken).ConfigureAwait(false);
+
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var syntaxFactsService = document.GetLanguageService<ISyntaxFactsService>();
+
+            var statementsToSurround = GetStatementsToSurround(declarationStatement, localVariable, semanticModel, syntaxFactsService, cancellationToken);
+
+            // Separate the newline from the trivia that is going on the using declaration line.
+            var trailingTrivia = SplitTrailingTrivia(declarationStatement, syntaxFactsService);
+
+            var usingStatement =
+                CreateUsingStatement(
+                    declarationStatement,
+                    trailingTrivia.sameLine,
+                    statementsToSurround)
+                    .WithLeadingTrivia(declarationStatement.GetLeadingTrivia())
+                    .WithTrailingTrivia(trailingTrivia.endOfLine);
+
+            if (statementsToSurround.Any())
+            {
+                var parentStatements = GetStatements(declarationStatement.Parent);
+                var declarationStatementIndex = parentStatements.IndexOf(declarationStatement);
+
+                var newParent = WithStatements(
+                    declarationStatement.Parent,
+                    new SyntaxList<TStatementSyntax>(
+                        parentStatements.Take(declarationStatementIndex)
+                        .Concat(new[] { usingStatement })
+                        .Concat(parentStatements.Skip(declarationStatementIndex + 1 + statementsToSurround.Count))));
+
+                return document.WithSyntaxRoot(root.ReplaceNode(declarationStatement.Parent, newParent
+                    .WithAdditionalAnnotations(Formatter.Annotation)));
+            }
+            else
+            {
+                return document.WithSyntaxRoot(root.ReplaceNode(declarationStatement, usingStatement
+                    .WithAdditionalAnnotations(Formatter.Annotation)));
```

Comment:
i'm finding a lot of the wrapping strategy confusing.  by breaking on the dot, it becomes less clear what this belongs to (indeed, i thought the .WithAdditionalAnnotations belonged to something else).  Can you do something like:

```c#
document.WithSyntaxRoot(root.ReplaceNode(
    declarationStatement,
    usingStatement.WithAdditionalAnnotations(Formatter.Annotation)));
```

benefits here are that siblings are aligned together (declarationStatement, and usingStatement.With...), and leaf expressions are not split.